shaders/src/fog.fs

```diff
@@ -0,0 +1,43 @@
+//------------------------------------------------------------------------------
+// Fog
+//------------------------------------------------------------------------------
+
+vec3 fog(vec3 color, vec3 cameraToWorld, vec3 wsCameraPosition, vec3 up) {
+    if (frameUniforms.fogDensity > 0.0) {
+        float wsCameraHeight = dot(up, wsCameraPosition);
+
+        // clamp to camera far plane so the skybox participates to the fog
+        // NOTE: the double-min is needed here, or some compilers won't work, apparently, it's
+        // because cameraToWorld may have infinites, min() with a non-constant, just doesn't work.
+        float cameraToWorldDistance = min(frameUniforms.cameraFar, min(MEDIUMP_FLT_MAX, length(cameraToWorld)));
+
+        // fog density with respect to camera's altitude
+        float density = frameUniforms.fogDensity * exp2(-frameUniforms.fogHeightFalloff * (wsCameraHeight - frameUniforms.fogHeight));
+
+        // falloff due to height of the shaded fragment w.r.t. the camera
+        float h = dot(cameraToWorld, up);
+        h = abs(h) > 0.01 ? h : 0.01;
+        float falloff = max(frameUniforms.fogHeightFalloff * h, -127.0);
+
+        // fog density at shaded fragment w.r.t. camera's altitude
+        float densityFalloff = density * (1.0 - exp2(-falloff)) / falloff;
+
+        // take fog start distance into account
+        float lineIntegral = densityFalloff * max(cameraToWorldDistance - frameUniforms.fogStart, 0.0);
+        float fogFactor = max(clamp(exp2(-lineIntegral), 0.0, 1.0), frameUniforms.fogMinOpacity);
```

Comment:
I'm not sure minOpacity is the right name for this. Since we multiply color by fogFactor, wouldn't it be the inverse of fog opacity?

Reply:
Yeah, I didn't think about it, this was from the prototype's code. A value of 0 means the fog can be fully opaque. We could call it `maxOpacity` and I could change the code so that a value of 1.0 is the default (fully opaque). This way you'd set it to like 0.9 if you wanted to limit the opacity of the fog.